### ISSUE TITLE
forward_confidence: the bounded score form is unusable at nap=8 (18.4x forward attenuation) — fixed by a normalised geometric-mean form

### ISSUE THREAD

**Issue by spikyclaudebot** [anatoli-starostin/spiky#112]
The LookupFFN-style forward-confidence gate multiplies each gathered LUT row by a smooth
per-(token, table) score derived from the routing margins `d = x[anchor_a] - x[anchor_b]`.
Its default form is a **product over `nap` factors**:

```
bounded (default):  score = prod_j sigmoid(2|d_j|)
margin           :  score = (sum_j |d_j|) * prod_j sigmoid(2|d_j|)     # the LookupFFN form
```

At the CompressionMHL anchor sizing (`nap = 8`, `H = 4`, `tph = 256`, `d_in = 32`,
`d_out = 48`) this default is **unusable**, and the reason is structural rather than a
coding error.

## Measured, on real activations

6,291,456 real margin vectors captured at the anchor sizing (`dump_margins.py`):

```
|d| margins    p25 0.179   median 0.381   p75 0.656   mean 0.454
score bounded  p25 0.0326  median 0.0465  p75 0.0670  mean 0.0542
score margin   p25 0.0943  median 0.165   p75 0.286   mean 0.229
```

Each factor is `sigmoid(2 * 0.38) ~ 0.68`, and `0.68^8 ~ 0.05`. **Every gathered table row
is multiplied by ~0.054 — an 18.4x forward attenuation.** The gradient consequence,
measured with the gate on and off:

```
            grad tables   grad decompress   grad x   grad compress
bounded        0.052x         0.051x        1.044x      1.045x
margin         0.262x         0.250x        1.332x      1.342x
```

The gate divides the learning signal reaching the LUT tables by ~19x while leaving the
routing gradients untouched.

## What it did to a real run

`exp_n_0178_A_fwdconf_seed1` (anchor config, `bounded`, seed 1) against its paired baseline
`sweep_s05_dout48_H4_tph256_c256_din32` (seed 1, gate off, final 1.434572):

| step | bounded gate | baseline S5 | gap |
|---|---|---|---|
| 500 | 1.987700 | 1.852331 | +0.135369 |
| 1000 | 1.868186 | 1.693385 | +0.174801 |
| 1500 | 1.804922 | 1.607591 | +0.197331 |
| 2000 | 1.753607 | 1.548721 | +0.204886 |
| 2500 | 1.718596 | 1.501781 | +0.216815 |
| 3000 | 1.692834 | 1.467895 | +0.224939 |

The gap **widens monotonically**; the run was stopped at step 3,200.

## Not a bug — a bad default

Train and eval are bit-identical under the gate (max delta 0), the native CUDA hard-eval
kernel is correctly disabled by `and not self.forward_confidence`, and no path
double-applies or skips the score. The implementation does exactly what it documents. What
is wrong is the *choice* of a product form at `nap = 8`: its attenuation compounds with the
anchor count. Measured across `nap`, on the empirical margin distribution:

```
 nap      bounded    bounded_norm       margin
   1     0.692675        0.692675     0.355763
   2     0.480238        0.687666     0.493968
   4     0.230857        0.685143     0.475681
   8     0.053137        0.683605     0.218524
  16     0.002820        0.682954     0.023177
```

`bounded` falls by 246x between nap=1 and nap=16. `margin` is non-monotone (its `sum|d|`
factor grows linearly while `prob` decays geometrically), peaking around nap=2 and
collapsing after — so it is not a general fix either, merely a milder failure at this nap.

## Fix shipped

A third form, `confidence_form="bounded_norm"` — the geometric mean of the same per-anchor
sigmoids, `(prod_j sigmoid(2|d_j|)) ** (1/nap)`, computed as `exp(mean_j logsigmoid(2m_j))`
so `exp(2m)` is never built. Same ordering as `bounded` (a monotone transform of it), but
**nap-independent in scale**: 0.6927 at nap=1 to 0.6830 at nap=16, a 1.4% drift. Measured at
the anchor sizing it scores 0.684 mean with the table gradient at 0.671x of gate-off.

`bounded` was deliberately left untouched rather than redefined, so every earlier run stays
bit-reproducible.

## Open question this raises

The normalisation also **flattens the gate almost to a constant**. Selectivity, measured as
the p75/p25 ratio of the score and as the within-token coefficient of variation across the
256 tables that get summed (the only part of the variance that actually reweights the
ensemble — the across-token part just rescales that token's FFN output):

```
                 p75/p25   within-token CV
bounded             2.06        0.536
margin              3.03        0.870
bounded_norm        1.09        0.061
```

A gate with p75/p25 = 1.09 is close to a constant multiplier, and a constant multiplier is
exactly absorbable into the linear, zero-initialised `decompress` that follows it — i.e. it
changes nothing the model could not already express. So if `bounded_norm` trains at parity
with the baseline, the honest reading is that **the gate is neutral here**, not that it
helps. Separating "the gate's scale was wrong" from "the gate's selectivity is useless"
needs a form with `bounded`'s spread and a corrected scale (a constant gain on `bounded`),
which is the natural next arm.

Sharpening the geometric mean with a temperature does **not** recover selectivity — measured,
`exp(mean_j logsigmoid(2*beta*m_j))` peaks at CV 0.073 around beta=2 and falls again, because
every factor saturates toward 1. That candidate is triaged out without spending GPU on it.

## Where the code and evidence live

Branch `research/ffn_replacement_fix`, local commits (not pushed):

- `30fc396e` — arm A stop, `STOPPED.md` with the curve, `diag_confidence_gate.py`,
  `diag_confidence_backward.py`, `gate_diagnosis.png`
- `f1276ec4` — `bounded_norm` implementation, its analytic backward, and tests
  (the derivative is the exact derivative of the normalised score, with a test asserting it
  is NOT the bounded derivative divided by nap); 347 lutorch tests pass
- `dump_margins.py` / `diag_confidence_forms.py` — the margin capture and the form triage
  that produced the nap and selectivity tables above

**Comment by spikyclaudebot** [comment]
## Resolved, plus a lead worth following

Five arms at the anchor sizing, all seed 1, all against baseline S5 = 1.434572 (vanilla-dense
zero-line S0 = 1.474749, cross-seed sd = 0.0096).

| arm | form | score mean | within-token CV | final | vs S5 |
|---|---|---|---|---|---|
| **D** | `margin` × 2.99 | 0.6836 | 0.946 | **1.432430** | **−0.002142** |
| **C** | `bounded` × 12.61 | 0.6839 | 0.584 | **1.434988** | +0.000416 |
| — | baseline S5, gate off | — | — | 1.434572 | 0 |
| **A′** | `bounded_norm` | 0.6838 | 0.067 | 1.441122 | +0.006550 |
| — | vanilla dense S0 | — | — | 1.474749 | +0.040177 |
| **B** | Light + `bounded_norm` | 0.6838 | 0.067 | 1.477708 | +0.043136 |
| **A** | `bounded` (this issue) | 0.0542 | 0.584 | stopped @3,200 | +0.225, widening |

### The original defect is settled: it was scale, not form

**Arm C is the proof.** The *identical* score shape that diverged to +0.225 as arm A lands
**+0.0004** from the baseline once multiplied by a constant. Nothing about `bounded`'s shape
was wrong — only its magnitude at nap=8.

It also **could not have healed itself**: margins widen only 1.69× over a full run, easing
bounded's attenuation from 18.4× to 7.7× and stopping there — and that is measured on the
*gate-off* trajectory, i.e. the optimistic case. "Give it more steps" is retired.

### Two things shipped

- **`confidence_form="bounded_norm"`** — the geometric mean, nap-invariant (0.6927 at nap=1 →
  0.6830 at nap=16 vs `bounded`'s 246× collapse). `bounded` was left untouched, so all earlier
  runs stay bit-reproducible.
- **`confidence_gain`** — a constant multiplier separating the gate's *scale* from its
  *selectivity*. Default 1.0 is bit-identical to no gain, forward and backward. This is what
  made arms C and D expressible at all; without it "the scale was wrong" and "the selectivity
  is useless" cannot be told apart.

361 lutorch tests pass.

### The lead: selectivity looks monotonically good

A′, C and D hold the score's mean fixed at 0.6838 and vary only its spread. The result is
monotone — more selective is better, across a 14× range of within-token CV, spanning 0.0087
bpb — and D is the only arm in the line that finishes *below* the baseline.

**This is not a result and should not be cited as one.** The spread (0.0087) is just under the
cross-seed sd (0.0096); a monotone ordering of three points has probability 1/6 under a
pure-noise null; and D carries its own confound (`margin` self-normalises during training —
0.229 → 0.944 — while its gain stays fixed, so part of its edge may be scale). The one
mitigating argument, which cuts both ways: these arms are *paired* on seed, `lut_base_seed`
and data order, whereas 0.0096 was measured across re-drawn inits, so the right denominator
may be smaller — but paired noise has never been measured here, so it can be used neither to
dismiss the trend nor to claim it.

**Follow-up that would settle it: 3 seeds each of A′ and D** (~4 GPU-hours), the two ends of
the axis.

### Separate finding: Light's detached routing costs +0.043 bpb

Arm B — `LightMultiHeadLUT`, routing address detached, so x gets gradient only through the
score — lands **+0.043136**, which is 4.5× the seed sd and therefore a *real* regression, and
marginally worse than vanilla dense. Predicted before the run from
`cos(light, fast) = +0.576` on `compress.weight`: aligned but not a rescaled copy, so it
should train (unlike arm A) and lag (unlike A′). It did both. Table and decompress gradients
are *identical* to Fast's — the handicap is confined to the input side.

### Recommendation

If the gate is used, `confidence_gain` is mandatory rather than optional, and `margin` is the
form to reach for (sharpest, and what the paper uses). `bounded_norm` is the safest default
(nap-invariant, no gain needed) but is the flattest and lands furthest behind. **`bounded`
without a gain should arguably stop being a default.**

Full write-up: `experiments/ffn_replacement/runs_corrected/LOOKUPFFN_LINE.md`, with the
figure alongside it. Branch `research/ffn_replacement_fix`, 15 local commits, not pushed.

**Comment by spikyclaudebot** [comment]
## Kernel fusion: Light reaches eval parity with Fast

The native eval kernel gathered each anchor pair, decided its sign bit, and discarded the
margin. A gated layer therefore had to gather every margin a **second** time in torch to
compute a per-table scalar it then reduced away — which is why Fast disables its native path
when the gate is on, and why Light's eval forward sat at 2.61 ms against Fast's 0.958 ms.

New kernel **`anchor_pairs_lookup_eval_forward_msb_scored`** returns `(indices, score)` from
one pass, accumulating the score from margins already in registers.

### The score is essentially free

```
native plain  (index only)          0.115 ms
native SCORED (index + score)       0.121 ms      +0.006 ms for the score
torch margin gather                 0.651 ms  }   both replaced:
torch score from those margins      1.019 ms  }   1.67 ms -> 0.12 ms  = 13.8x
```

Folding the margin gather in was not a separate step — it is the same step. The margins never
leave registers, so the 0.65 ms does not shrink, it leaves the eval path entirely.

### Numerics vs the torch path

All three forms, gains 1.0 and 12.61:

| | index | score max&#124;rel&#124; |
|---|---|---|
| float64 | bit-identical | 4.4e-16 … 2.3e-15 |
| float32 | bit-identical | 2.6e-07 … 1.1e-06 |

`m = |d| >= 0`, so `exp(-2m) <= 1` and `log1p(exp(-2m))` is evaluated directly — `exp(2m)` is
never materialised, the same discipline as the torch path, which is what lets the two agree
to round-off rather than merely closely.

### Result

Layer, fp32, 6,144 tokens:

| | fwd(eval) | fwd(train) | fwd+bwd | bwd only | peak MiB |
|---|---|---|---|---|---|
| fast, gate off | 0.958 | 1.015 | 34.013 | 32.998 | 19,949 |
| fast + bounded_norm | 1.142 | 1.164 | 44.718 | 43.554 | 21,368 |
| light + bounded_norm | **1.043** | 2.691 | **9.090** | **6.399** | **996** |

Whole model (~105M): eval forward **11.1 ms** (Fast 10.3, gated Fast 11.6, dense 6.9); train
step **68.4 ms** (Fast 218.0, gated Fast 282.1, dense 19.8); peak **5.51 GiB** vs Fast's 21.4.

Eval forward went 8.51 → 2.69 (embedding_bag fusion) → **0.96 ms**. That is 1.09× of Fast's
native ungated path and **faster than gated Fast**.

### Iso-inference-cost, recomputed

Using the measured eval exponent (time ~ tph^1.19) and the budget law's −0.00746 bpb/doubling:

| | ratio | budget given back | Light at iso-cost |
|---|---|---|---|
| vs gate-off Fast | 1.08× dearer | +0.0007 bpb | 1.4784 vs 1.4346 (**+0.0438**) |
| vs vanilla dense | 1.61× dearer | +0.0043 bpb | 1.4820 vs 1.4747 (**+0.0073**) |

+0.0073 is **inside the 0.0096 seed sd**, so at equal inference cost Light is statistically
indistinguishable from dense. And the remaining +0.0438 gap to Fast is now **quality, not
speed** — it is Fast's directional surrogate that earns it, not its kernel.

(The budget-law slope was fitted on 16k runs; applying it to these 4k proxies is an
extrapolation.)

### Scope — the fused path is eval-only, structurally

Autograd needs the margins as a differentiable tensor to reach `x` through the score, and this
path deliberately never materialises them. A guard disables it whenever grad is enabled, with
a test asserting the grad-enabled forward still carries its graph and produces a nonzero
`grad_x`. Training stays on the torch path — which is also the right call on the numbers:
fusing there would buy ~1.12× (forward is 2.7 ms of a 9.1 ms fwd+bwd) against ~2.7× here.

### We are now at the gather floor

Of the 0.961 ms forward: native address+score **0.121 ms (13%)**, `embedding_bag`
**0.783 ms (81%)** — one random row read per (token, table), within 1.23× of irreducible.
Further kernel work on the score buys nothing.

Remaining headroom is algorithmic only, and noted rather than pursued: fewer tables per token
or narrower rows (both trade directly against table budget, so the budget law taxes them);
int8 tables as a **cache-residency** bet rather than a bandwidth one (bf16 already showed
bandwidth is not the constraint — it halved table bytes and changed the forward by <2%); and
address-packing locality so a token's 256 reads touch fewer cache lines.

### Build note

The `.so` is an untracked build artefact. Rebuilding needs `/usr/local/cuda-13.3` and
`TORCH_CUDA_ARCH_LIST=12.0`, then the `.so` copied into the venv. Without a rebuild the scored
kernel is **silently absent** and the layer falls back to torch — correct but slow. The tests
skip rather than fail in that case, deliberately, so a machine without the extension does not
show red.

**397 lutorch tests pass.** 18 local commits on `research/ffn_replacement_fix`, nothing pushed,
no new training runs.

---

### Erratum, recorded here because it will become visible on push

Two commits in this stack cite **#111** in their subject lines when they mean **#112**:

- `30fc396e` — "arm A stopped at step 3,200: the bounded confidence gate attenuates 18x…"
- `f1276ec4` — "confidence_form \"bounded_norm\": the geometric mean of the per-anchor sigmoids"

#111 is "Distill LUTs from a pretrained FFN", unrelated. `5eb9bcec` records the correction and
fixes the in-source comments; the two subjects were left alone rather than rewriting history.
Nothing is pushed yet, so GitHub has not cross-linked them anywhere — but **once this branch is
pushed, both commits will appear on #111**. Worth a glance from whoever pushes: either accept
the noise, or squash/reword those two subjects before the push. I have not touched #111, since
a note there would be premature while no link exists.
